staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -545,3 +571,138 @@ func (a *Authenticator) wrapCmdRunErrorLocked(err error) error {
 		return fmt.Errorf("exec: %v", err)
 	}
 }
+
+// `updateCommandAndCheckAllowlistLocked` determines whether or not the specified executable may run
+// according to the credential plugin policy. If the plugin is allowed, `nil`
+// is returned. If the plugin is not allowed, an error must be returned
+// explaining why.
+func (a *Authenticator) updateCommandAndCheckAllowlistLocked(cmd *exec.Cmd) error {
+	switch a.execPluginPolicy.PolicyType {
+	case "", api.PluginPolicyAllowAll:
+		return nil
+	case api.PluginPolicyDenyAll:
+		return fmt.Errorf("plugin %q not allowed: policy set to %q", a.cmd, api.PluginPolicyDenyAll)
+	case api.PluginPolicyAllowlist:
+		return a.checkAllowlistLocked(cmd)
+	default:
+		return fmt.Errorf("unknown plugin policy %q", a.execPluginPolicy.PolicyType)
+	}
+}
+
+// `checkAllowlistLocked` checks the specified plugin against the allowlist,
+// and may update the Authenticator's allowlistLookup set.
+func (a *Authenticator) checkAllowlistLocked(cmd *exec.Cmd) error {
+	if cmd.Err != nil {
+		// we know this will fail execution, so bail early
+		return fmt.Errorf("plugin command has error after initialization: %w", cmd.Err)
+	}
+
+	// cmd.Path will by this point be either an absolute path or a relative
+	// path with at least one path separator. a.cmd may be a basename. If there
+	// is an exact match in the allowlist, return success.
+	if a.allowlistLookup.Has(a.cmd) || a.allowlistLookup.Has(cmd.Path) {
+		return nil
+	}
```

Comment:
code-reviewed in person, pushed comments / simplifications to https://github.com/liggitt/kubernetes/commits/pmengelbert/kuberc/4/ if you want to pull in the last commit (can squash in)